services/ml/compilation_delegate_dnnl.cc

```diff
@@ -0,0 +1,1764 @@
+    *dnnl_type = dnnl_f32;
+  } else if (type == mojom::TENSOR_INT32) {
+    *dnnl_type = dnnl_s32;
+  } else {
+    LOG(ERROR) << "Type " << type << " is not supported";
+    return mojom::BAD_DATA;
+  }
+  return mojom::NOT_ERROR;
+}
+
+int32_t CompilationDelegateDnnl::DnnlCreateMemoryByQueryType(
+    const dnnl_primitive_desc_t& pd,
+    dnnl_query_t query_type,
+    dnnl_memory_t& output_memory) {
+  const dnnl_memory_desc_t* output_desc =
+      LATE(dnnl_primitive_desc_query_md)(pd, query_type, 0);
+  dnnl_status_t status = LATE(dnnl_memory_create)(
+      &output_memory, output_desc, compiled_model_->engine, NULL);
+  if (status != dnnl_success) {
+    LOG(ERROR) << "[DNNL] failed to create memory " << status;
+    return mojom::OP_FAILED;
+  }
+  size_t buffer_size = LATE(dnnl_memory_desc_get_size)(output_desc);
+  void* buffer = base::AlignedAlloc(buffer_size, ALIGNMENT);
+  status = LATE(dnnl_memory_set_data_handle)(output_memory, buffer);
+  if (status != dnnl_success) {
+    LOG(ERROR) << "[DNNL] failed to set data handle to memory primitive "
+               << status;
+    LATE(dnnl_memory_destroy)(output_memory);
+    base::AlignedFree(buffer);
+    return mojom::OP_FAILED;
+  }
+  return mojom::NOT_ERROR;
+}
+
+int32_t CompilationDelegateDnnl::DnnlCreateMemoryDescriptor(
+    int32_t index,
+    dnnl_memory_desc_t& memory_desc,
+    dnnl_format_tag_t* user_format) {
+  const mojom::ModelInfoPtr& model = compilation_->GetModel();
+  const mojom::OperandPtr& operand = model->operands[index];
+  dnnl_data_type_t data_type;
+  int32_t result = DnnlGetDataType(operand->type, &data_type);
+  if (result != mojom::NOT_ERROR) {
+    return result;
+  }
+  dnnl_format_tag_t format;
+  if (!user_format) {
+    result = DnnlGetMemoryFormat(operand->dimensions, &format);
+    if (result != mojom::NOT_ERROR) {
+      return result;
+    }
+  } else {
+    format = *user_format;
+  }
+
+  std::vector<dnnl_dim_t> dims;
+  result = DnnlGetDims(operand->dimensions, dims, format);
+  if (result != mojom::NOT_ERROR) {
+    return result;
+  }
+
+  dnnl_status_t status = LATE(dnnl_memory_desc_init_by_tag)(
+      &memory_desc, dims.size(), dims.data(), data_type, format);
+  if (status != dnnl_success) {
+    LOG(ERROR) << "[DNNL] failed to init memory descriptor " << status;
+    return mojom::OP_FAILED;
+  }
+  return mojom::NOT_ERROR;
+}
+
+int32_t CompilationDelegateDnnl::DnnlCreateMemory(
+    uint32_t index,
+    dnnl_memory_t& memory,
+    dnnl_format_tag_t* user_format) {
+  dnnl_memory_desc_t memory_desc;
+  int32_t result = DnnlCreateMemoryDescriptor(index, memory_desc, user_format);
+  if (result != mojom::NOT_ERROR) {
+    return result;
+  }
+
+  dnnl_status_t status = LATE(dnnl_memory_create)(
+      &memory, &memory_desc, compiled_model_->engine, DNNL_MEMORY_ALLOCATE);
+  if (status != dnnl_success) {
+    LOG(ERROR) << "[DNNL] failed to create memory primitive " << status;
+    status = LATE(dnnl_memory_destroy)(memory);
```

Comment:
The memory doesn't destroy if it's created unsuccessfully?

Reply:
It doesn't need to destory if it's created unsuccessfully.